azkaban-common/src/main/java/azkaban/utils/RestfulApiClient.java

```diff
@@ -0,0 +1,239 @@
+   * @param uri   the URI of the request.
+   * @param headerEntries   extra entries to be added to request header.
+   * @param postingBody  the content to be posted , optional.
+   * @return the response object type of which is specified by user.
+   * @throws UnsupportedEncodingException, IOException */
+  public T httpPut(URI uri, List<NameValuePair> headerEntries,
+      String postingBody) throws UnsupportedEncodingException, IOException{
+    // shortcut if the passed url is invalid.
+    if (null == uri){
+      logger.error(" unable to perform httpPut as the passed url is null or empty.");
+      return null;
+    }
+
+    HttpPut put = new HttpPut(uri);
+    return this.sendAndReturn(completeRequest(put, headerEntries, postingBody));
+  }
+
+  /** function to dispatch the request and pass back the response.
+   * */
+  protected T sendAndReturn(HttpUriRequest request) throws IOException{
+    CloseableHttpClient client = HttpClients.createDefault();
+    try {
+      return this.parseResponse(client.execute(request));
+    }finally{
+      client.close();
+    }
+  }
+
+  /** helper function to build a valid URI.
+   *  @param host   host name.
+   *  @param port   host port.
+   *  @param path   extra path after host.
+   *  @param isHttp indicates if whether Http or HTTPS should be used.
+   *  @param params extra query parameters.
+   *  @return the URI built from the inputs.
+   *  @throws IOException
+   * */
+  public static URI BuildUri(String host, int port, String path,
+      boolean isHttp, Pair<String, String>... params) throws IOException{
+    URIBuilder builder = new URIBuilder();
+    builder.setScheme(isHttp? "http" : "https").setHost(host).setPort(port);
+
+    if (null != path && path.length() > 0){
+      builder.setPath(path);
+    }
+
+    if (params != null) {
+      for (Pair<String, String> pair : params) {
+        builder.setParameter(pair.getFirst(), pair.getSecond());
+      }
+    }
+
+    URI uri = null;
+    try {
+      uri = builder.build();
+    } catch (URISyntaxException e) {
+      throw new IOException(e);
+    }
+
+    return uri;
+  }
+
+  /** helper function to build a valid URI.
+   *  @param uri    the URI to start with.
+   *  @param params extra query parameters to append.
+   *  @return the URI built from the inputs.
+   *  @throws IOException
+   * */
+  public static URI BuildUri(URI uri, Pair<String, String>... params) throws IOException{
+    URIBuilder builder = new URIBuilder(uri);
+
+    if (params != null) {
+      for (Pair<String, String> pair : params) {
+        builder.setParameter(pair.getFirst(), pair.getSecond());
+      }
+    }
+
+    URI returningUri = null;
+    try {
+      returningUri = builder.build();
+    } catch (URISyntaxException e) {
+      throw new IOException(e);
+    }
+
+    return returningUri;
+  }
+
+  /** helper function to fill  the request with header entries .
+   * */
+  private static HttpMessage completeRequest(HttpMessage request,
```

Comment:
OK.  However the name "completeRequest" doesn't tell the reader much.  Since it is a private method, then it doesn't matter too much.